### PR TITLE
Feature/time triggered oracle source for trading terminated

### PR DESCRIPTION
**Pull request by guoguojin**
Closes #4478

### CODE REVIEW

**Review by ValentinTrinque — CHANGES_REQUESTED** [review]

**Review by jeremyletang — CHANGES_REQUESTED** [review]
Just requesting change to block the merge as I want to go through it before it's merged, and don't have the time just now.

**Review by ValentinTrinque — CHANGES_REQUESTED** [review]

**Review by jeremyletang — CHANGES_REQUESTED** [review]

**Review by wwestgarth — DISMISSED** [review]

**Review by ValentinTrinque — CHANGES_REQUESTED** [review]

**Review by ValentinTrinque — CHANGES_REQUESTED** [review]

**Review by ValentinTrinque — CHANGES_REQUESTED** [review]

**Review by ValentinTrinque — CHANGES_REQUESTED** [review]

**Review by ValentinTrinque — APPROVED** [review]

**Review by jeremyletang — APPROVED** [review]
Looks good, might need a rebase though now.

**Pull request merged**
